### DISCUSSION TITLE
Specification of air as the ONLY working fluid - massfrac

### DISCUSSION THREAD

**Discussion by SRkumar97** [AMReX-Combustion/PeleC#868] [Q&A] [answered]
Hi,
I am trying to run a supersonic boundary layer case over a flat plate, with standard air as the gas.
Now the issue I am facing is wrt the mass fractions and gas_IDs of N2 and O2. I want specify clearly that there is only and only air as the gas. 
Now, I found a case which has got N2 and O2 with mass fractions as that of standard air - the Spray-Multijet testcase. Here, the prob_parm.H has the mass fractions stored in the variables Y_N2 and Y_O2 respectively. And the **prob.cpp** and **prob.H** files call for these two variables in massfrac[N2_ID] and massfrac[O2_ID] (e.g. - **Spray-Multijet** and **Spray-Conv**)

My doubt is, whether these two IDs are pre-defined in PeleC? In certain cases I'm seeing that an additional mass fraction is also specified with a value of absolute 0.0, (e.g. **amrex::Real massfrac[NUM_SPECIES] = {0.0};** in **Spray-TG**) . But I don't understand this additional statement. I am unable to find any value for NUM_SPECIES in the prob_parm.H.

Would be glad if someone could explain all this and inform what is the correct way to specify air as the only working fluid.

I also came across this case involving air -

https://amrex-combustion.github.io/PeleC/validation/validation.html#counter-flow-diffusion-flame

But I am unable to find a case with a matching name in the RegTests folder.

Thanks for reading!

**Comment by d-montgomery** [comment, marked as answer]
PelePhysics offers several chemistry mechanisms that include O2. In the GNUmakefile for the Spray-Multijet case, you'll see the line `Chemistry_Model := decane_3sp`, which points to [this specific mechanism](https://github.com/AMReX-Combustion/PelePhysics/tree/development/Mechanisms/decane_3sp) in PelePhysics.  In the associated `mechanism.H` file, you'll find three species defined: NC10H22, O2, and N2, each with a pre-defined index ID for the mass fraction.  This file also contains the definition for `NUM_SPECIES`.

If you want to run a case with [air](https://github.com/AMReX-Combustion/PelePhysics/tree/development/Mechanisms/air), you can set `Chemistry_Model := air`, then initialize the mass fraction in PeleC using the pre-defined index ID's in the associated `mechanism.H` file.  The `air` mechanism has two species, O2 and N2.  If you only want O2, initialize `massfrac[O2_ID] = 1.0` and `massfrac[N2_ID] = 0.0`, then use corresponding consistent boundary conditions.

**Reply by SRkumar97** [reply]
Hi, I have one more point to clarify here. I am seeing that in cases like **EB-InflowBC** wherein we are using N2 and O2 separately by assigning the massfrac IDs, the massfractions of left and right side gases (N2 and O2 resp.) have been set to {0.0} and then to 1.0 in lines 38-41 of the **prob.cpp** as pasted below.

  amrex::Real massfrac_l[NUM_SPECIES] = {0.0};
  amrex::Real massfrac_r[NUM_SPECIES] = {0.0};
  massfrac_l[PeleC::h_prob_parm_device->left_gas_id] = 1.0;
  massfrac_r[PeleC::h_prob_parm_device->right_gas_id] = 1.0;

I do not understand why we are setting a value of 0.0 first, using the [NUM_SPECIES] array. Could you please clarify?

**Reply by d-montgomery** [reply]
The EB-InflowBC case injects N2 into a domain initially filled with O2.  The left gas is O2 and the injected right gas is N2.  The lines of code following the ones you are questioning in the `prob.cpp` use the PelePhysics equations of state (specifically `eos.PYT2RE`) to get the density, R, and the energy, E, of the left and right gases , respectively.

`eos.PYT2RE` requires the argument `Y` to be an `amrex::Real massfrac[NUM_SPECIES]`, which in this case is an array with 2 elements.  Before the `eos` calls, the left and right mass fractions are initialized to ensure that all elements are set to zero, and then one of the elements is updated to 1.0. Initializing variables in C++ is considered best practice, as failing to do so can lead to undesirable or undefined quantities.

**Reply by SRkumar97** [reply]
Ok! So in a case wherein there is no usage of eos(), I can skip these two lines of initialization for N2 and O2 (amrex::Real massfrac_l[NUM_SPECIES] = {0.0}; **and** amrex::Real massfrac_r[NUM_SPECIES] = {0.0};)
And type directly the massfractions of N2 and O2, to be read from the prob_parm.H file, correct?
e.g.
  massfrac[O2_ID] = prob_parm.massfrac_O2;
  massfrac[N2_ID] = prob_parm.massfrac_N2;

**Reply by d-montgomery** [reply]
Yes, it is fine to manually initialize each species to a known value.  Just be sure to initialize every element of the massfrac.  If you select a mechanism with multiple species, but are only using a couple of species, initialize every element to zero first.  I hope this helps.

**Reply by SRkumar97** [reply]
Got it, thanks!

**Reply by SRkumar97** [reply]
Hi, I was able to do my shock simulation, under inviscid assumption using Gamma Law, which is the ideal gas equation for standard air, with R=287 J/Kg-K. Since my case has no kind of reactions and is standard air, Gamma Law works fine. Now I want to run a supersonic boundary layer case. Here, I am happy and well set with using the Gamma Law and Sutherland's law for dynamic viscosity. I am trying to run standard boundary layer flows like in other CFD solvers. However, PeleC doesn't seem to allow this combination of Gamma Law with Sutherland's law. The exec compilation throws an error for this combination. How do I use Gamma Law with Sutherland's law, then? 

**Note** - Here I'm only interested in Sutherland's law for dynamic viscosity. Which will reflect in the values of shear stress. Rest of the transport coefficients are unimportant; can be assumed constant.

**Reply by d-montgomery** [reply]
Hi @SRkumar97, have you been able to figure out a work around for your problem yet?  PeleC calls [PelePhysics for the equations of state](https://amrex-combustion.github.io/PelePhysics/EOS.html), which includes `GammaLaw`, `Fuego`, and `Soave-Redlich-Kwong`, but does not include Sutherland's law.  

Will dynamic viscosity be constant in your case? If so, one option is to use a `Null` chemistry model and manually set the constant transport coefficients in `prob.cpp`.  In the `GNUmakefile` set:
~~~
Eos_Model := GammaLaw
Chemistry_Model := Null
Transport_Model := Constant
~~~
Then, in `prob.cpp` set the dynamic viscosity using Sutherland's law similar to [what is done here](https://github.com/dmontgomeryNREL/PeleC/blob/4ba05a820a73cb06d81bdd681b313b2ab1a218ac/Exec/RegTests/EB-FlowPastCylinder/prob.cpp#L53) in the `FlowPastCylinder` regular test.

**Reply by baperry2** [reply]
Adding to what @dmontgomeryNREL said, PelePhysics does support `Sutherland` for the transport model natively, but only when the `Eos_Model` is set to `Fuego`. So for air-only simulations, you could use
```
Eos_Model := Fuego
Chemistry_Model := air
Transport_Model := Sutherland
```
which would be slightly more expensive but have slightly more accurate thermodynamics than using `GammaLaw`. 

The reason we don't support the combination of Sutherland and GammaLaw natively is the [implementation of Sutherland transport involves the CKCPBS function](https://github.com/AMReX-Combustion/PelePhysics/blob/66113ee077810253e10a4b891bb4911706f59471/Source/Transport/Sutherland.H#L45) from the Chemistry_Model, which is not implemented for Null chemistry. You could use the native implementation if you replace `CKCPBS` with `eos.TY2Cp` and then delete the static assert that raises the error. If you do this consider submitting a PR with the change. See #384.

**Reply by SRkumar97** [reply]
Thank you very much @baperry2 and @dmontgomeryNREL !

@dmontgomeryNREL I was able to figure out a way, however not exactly the combination I need. Yes, I have decided to use Fuego Law with Transport model as Simple/Sutherland law. However this gives the combination of a thermally perfect gas and varying dynamic viscosity. I am okay to work with, but need to be careful with all BCs and boundary layer profiles (turbinflow).

Yes, @baperry2 I found this way to be computationally more expensive. Since my objective is to simply run LES/Hybrid LES/DNS of non-reacting, general compressible flows, the calorically perfect gas assumption along with varying dynamic viscosity is the perfect combination I need. 
As you have put forward, I'll consider editing this function and submit a PR here, if I am successful. Thanks a ton, once again!

> Adding to what @dmontgomeryNREL said, PelePhysics does support `Sutherland` for the transport model natively, but only when the `Eos_Model` is set to `Fuego`. So for air-only simulations, you could use
> 
> ```
> Eos_Model := Fuego
> Chemistry_Model := air
> Transport_Model := Sutherland
> ```
> 
> which would be slightly more expensive but have slightly more accurate thermodynamics than using `GammaLaw`.
> 
> The reason we don't support the combination of Sutherland and GammaLaw natively is the [implementation of Sutherland transport involves the CKCPBS function](https://github.com/AMReX-Combustion/PelePhysics/blob/66113ee077810253e10a4b891bb4911706f59471/Source/Transport/Sutherland.H#L45) from the Chemistry_Model, which is not implemented for Null chemistry. You could use the native implementation if you replace `CKCPBS` with `eos.TY2Cp` and then delete the static assert that raises the error. If you do this consider submitting a PR with the change. See #384.

**Comment by SRkumar97** [comment]
Thank you very much, I am able to correlate and understand whatever you have explained through the suite of testcases.
